File: LeetCodeKotlin/JumpGameII_45.kt
class JumpGameII_45 {
    fun jump(nums: IntArray): Int {
        if (nums.size == 1) return 0

        var jumps = 0         // Count of jumps made.
        var currentEnd = 0    // Current window end.
        var farthest = 0      // Farthest reachable index so far.

        // Iterate over indices from 0 to the second-to-last element.
        for (i in 0 until nums.lastIndex) {
            // Update the farthest index we can reach.
            farthest = maxOf(farthest, i + nums[i])

            // When we've reached the end of the current window,
            // we make a jump and set the window's end to the farthest index reached.
            if (i == currentEnd) {
                jumps++
                currentEnd = farthest
            }
        }
        return jumps
    }
}

File: LeetCodeCpp/JumpGameII_45.cpp
#include "JumpGameII_45.h"
#include <algorithm>

int JumpGameII_45::jump(vector<int>& nums) {
    const int n = (int)nums.size();
    if (n <= 1) return 0;

    int jumps = 0;
    int curEnd = 0;    // farthest index reachable with exactly `jumps` jumps
    int farthest = 0;  // farthest index reachable with `jumps + 1` jumps

    for (int i = 0; i < n - 1; ++i) {
        farthest = max(farthest, i + nums[i]);
        if (i == curEnd) {      // end of current layer -> must "take" a jump
            ++jumps;
            curEnd = farthest;
        }
    }
    return jumps;
}
